### PR TITLE
Speedup Interflow

### PR DESCRIPTION
**Pull request by JD557**
Adds some performance optimizations to Interflow, mainly regarding hash map cloning and avoiding unnecessary operations when logging is disabled.

I ran this on a [small codebase](https://github.com/JD557/minart/blob/06ecc9edbe99ffaa221f6e239895085e81bf71fa/examples/colorsquare/shared/src/main/scala/eu/joaocosta/minart/examples/ColorSquare.scala) and got some nice performance improvements on `release-full`.

|  | Optimization (Avg) | Total (Avg) |
|-|-|-|
| 0.4.0 | 170934ms | 226639ms |
| This Branch | 90806ms | 156839ms |

It might be nice to also try this on a large codebase. Any ideas?

-----

Full times of the 3 runs (SBT was not restarted between runs to keep the JIT warm, it was restarted between tests):

# 0.4.0

## Run 1

```
[info] Linking (2373 ms)
[info] Discovered 1083 classes and 9147 methods
[info] Optimizing (release-full mode) (173719 ms)
[info] Generating intermediate code (7042 ms)
[info] Produced 8 files
[info] Compiling to native code (15208 ms)
[info] Linking native code (immix gc, thin lto) (33952 ms)
[info] Total (232533 ms)
```

## Run 2

```
[info] Linking (1116 ms)
[info] Discovered 1083 classes and 9147 methods
[info] Optimizing (release-full mode) (169880 ms)
[info] Generating intermediate code (5002 ms)
[info] Produced 8 files
[info] Compiling to native code (13307 ms)
[info] Linking native code (immix gc, thin lto) (34893 ms)
[info] Total (224436 ms)
```

## Run 3

```
[info] Linking (1167 ms)
[info] Discovered 1083 classes and 9147 methods
[info] Optimizing (release-full mode) (169202 ms)
[info] Generating intermediate code (4851 ms)
[info] Produced 8 files
[info] Compiling to native code (13107 ms)
[info] Linking native code (immix gc, thin lto) (34634 ms)
[info] Total (223112 ms)
```

# New Branch

## Run 1

```
[info] Linking (2868 ms)
[info] Discovered 1083 classes and 9147 methods
[info] Optimizing (release-full mode) (94776 ms)
[info] Generating intermediate code (8554 ms)
[info] Produced 8 files
[info] Compiling to native code (13660 ms)
[info] Linking native code (immix gc, thin lto) (45077 ms)
[info] Total (165277 ms)
```

## Run 2

```
[info] Linking (1186 ms)
[info] Discovered 1083 classes and 9147 methods
[info] Optimizing (release-full mode) (89203 ms)
[info] Generating intermediate code (5405 ms)
[info] Produced 8 files
[info] Compiling to native code (13693 ms)
[info] Linking native code (immix gc, thin lto) (45296 ms)
[info] Total (154986 ms)
```

## Run 3

```
[info] Linking (971 ms)
[info] Discovered 1083 classes and 9147 methods
[info] Optimizing (release-full mode) (88440 ms)
[info] Generating intermediate code (4288 ms)
[info] Produced 8 files
[info] Compiling to native code (12454 ms)
[info] Linking native code (immix gc, thin lto) (44035 ms)
[info] Total (150305 ms)
```

### CODE REVIEW

**Comment by JD557** [comment]
CLA signed, please re-run the check.

**Comment by ekrich** [comment]
Thank you for the PR. It looks like you need to run `scalafmt` on the code. Try `$ scripts/scalafmt` inside the `scala-native` directory.

File: tools/src/main/scala/scala/scalanative/interflow/Inline.scala [hunk 1]
```diff
@@ -20,17 +20,17 @@ trait Inline { self: Interflow =>
       .fold[Boolean] {
         false
       } { defn =>
-        val isCtor = originalName(name) match {
+        lazy val isCtor = originalName(name) match {
```
Comment (WojciechMazur) [inline tools/src/main/scala/scala/scalanative/interflow/Inline.scala:23]: This and following `lazy val`s can be potentially replaced with `def`s. They're used only once and `lazy val`s implies some amounts of overhead.

File: tools/src/main/scala/scala/scalanative/interflow/Inline.scala [hunk 2]
```diff
@@ -20,17 +20,17 @@ trait Inline { self: Interflow =>
...
           case Global.Member(_, sig) if sig.isCtor || sig.isImplCtor =>
             true
           case _ =>
             false
         }
-        val isSmall =
+        lazy val isSmall =
           defn.insts.size <= 8
         val isExtern =
```
Comment (WojciechMazur) [inline tools/src/main/scala/scala/scalanative/interflow/Inline.scala:31]: `isExtern` and other conditions can also be treated with replacement to `lazy val` / `def`. They're also used only once

**Review by WojciechMazur — CHANGES_REQUESTED** [review]
Overall it looks good, thank you for this optimization, the results look really promising, especially when most users tend to use `releaseFast` instead of `releaseFull` due amount of time it's taking. 
As @ekrich already suggested you'll need to run `scipts/scalafmt` from the project directory and push fixed formatting. 
Also since it's your first PR to Scala Native, I'd like to welcome on board!

File: tools/src/main/scala/scala/scalanative/interflow/MergeProcessor.scala [hunk 3]
```diff
@@ -101,8 +101,8 @@ final class MergeProcessor(insts: Array[Inst],
           def mergeLocal(local: Local, value: Val): Unit = {
             val values = mutable.UnrolledBuffer.empty[Val]
             states.foreach { s =>
-              if (s.locals.contains(local)) {
-                values += s.locals(local)
+              s.locals.get(local).foreach { l =>
```
Comment (WojciechMazur) [inline tools/src/main/scala/scala/scalanative/interflow/MergeProcessor.scala:104]: Also we can simply it the following:
```suggestion
              s.locals.get(local).foreach(values += _ )
```

File: tools/src/main/scala/scala/scalanative/interflow/Inline.scala (hunk 1 above)
Comment (JD557) [reply tools/src/main/scala/scala/scalanative/interflow/Inline.scala]: Changed in a1df6de.

File: tools/src/main/scala/scala/scalanative/interflow/Inline.scala (hunk 2 above)
Comment (JD557) [reply tools/src/main/scala/scala/scalanative/interflow/Inline.scala:31]: Changed in a1df6de.

I left a lazy val in `shallNot` to make sure that things don't get too slow when logging is enabled.

File: tools/src/main/scala/scala/scalanative/interflow/MergeProcessor.scala (hunk 3 above)
Comment (JD557) [reply tools/src/main/scala/scala/scalanative/interflow/MergeProcessor.scala]: Applied in 81ba80c.

**Comment by JD557** [comment]
>Thank you for the PR. It looks like you need to run scalafmt on the code. Try $ scripts/scalafmt inside the scala-native directory.

>As @ekrich already suggested you'll need to run scipts/scalafmt from the project directory and push fixed formatting.

Done in 73f3f53.

> Also since it's your first PR to Scala Native, I'd like to welcome on board!

Thank you :)

**Review by WojciechMazur — APPROVED** [review]

**Pull request merged**